### PR TITLE
[scripts-audit] vcpkg.cmake

### PR DESCRIPTION
**Pull request by strega-nil**
See PR #16055

This also (attempts to) allow vcpkg.cmake to run with cmake 3.1; we had a `VERSION_GREATER_EQUAL` check, but that if keyword was added in cmake 3.7.

### CODE REVIEW

File: scripts/buildsystems/vcpkg.cmake [hunk 1]
```diff
@@ -586,23 +578,56 @@ if(X_VCPKG_APPLOCAL_DEPS_INSTALL)
...
+function(foo_replacement)
+    z_vcpkg_escape_argv(ESCAPED_ARGV)
+    foo(${ESCAPED_ARGV})
+    ...
+endfunction()
+
+this macro allows a function to get a variable with its arguments,
+escaped correctly so that list arguments are not passed as splats to the underlying function.
+#]===]
+macro(z_vcpkg_escape_argv OUT)
```
Comment (BillyONeal) [inline scripts/buildsystems/vcpkg.cmake:592]: Does it matter that this is used before it is declared?

Comment (BillyONeal) [reply scripts/buildsystems/vcpkg.cmake:592]: Also I don't know how to audit this for correctness so other reviewers pay close attention.

Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:592]: 1. It doesn't but 2. I still think we should move z_vcpkg_escape_argv up.

File: scripts/buildsystems/vcpkg.cmake [hunk 2]
```diff
@@ -589,20 +731,29 @@ endif()
 if(NOT DEFINED VCPKG_OVERRIDE_FIND_PACKAGE_NAME)
     set(VCPKG_OVERRIDE_FIND_PACKAGE_NAME find_package)
 endif()
-macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
+function("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}")
```
Comment (Neumann-A) [inline scripts/buildsystems/vcpkg.cmake:734]: this potentially scopes the `find_package` call which is not intended

File: scripts/buildsystems/vcpkg.cmake [hunk 3]
```diff
@@ -217,76 +213,77 @@ else()
...
+        set(Z_VCPKG_INSTALLED_DIR "${X_VCPKG_INSTALLED_DIR}")
+    elseif(VCPKG_MANIFEST_MODE)
+        set(Z_VCPKG_INSTALLED_DIR "${CMAKE_BINARY_DIR}/vcpkg_installed")
     else()
-        set(_VCPKG_INSTALLED_DIR ${_VCPKG_ROOT_DIR}/installed)
+        set(Z_VCPKG_INSTALLED_DIR "${Z_VCPKG_ROOT_DIR}/installed")
     endif()
 
-    set(_VCPKG_INSTALLED_DIR ${_VCPKG_INSTALLED_DIR}
+    set(Z_VCPKG_INSTALLED_DIR "${Z_VCPKG_INSTALLED_DIR}"
```
Comment (Neumann-A) [inline scripts/buildsystems/vcpkg.cmake:258]: changing _VCPKG_INSTALLED_DIR probably requires replacement in a lot of vcpkg-cmake-wrapper.cmake

File: scripts/buildsystems/vcpkg.cmake [hunk 4]
```diff
@@ -589,20 +731,29 @@ endif()
...
     endif()
-    string(TOLOWER "${name}" _vcpkg_lowercase_name)
 
-    if(EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake")
-        set(ARGS "${ARGV}")
-        include(${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake)
-    elseif("${name}" STREQUAL "Boost" AND EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include/boost")
+    z_vcpkg_escape_argv(ESCAPED_ARGV)
+    set(TARGET_NAME "${ARGV0}")
+    string(TOLOWER "${TARGET_NAME}" LOWERCASE_TARGET_NAME)
```
Comment (Neumann-A) [inline scripts/buildsystems/vcpkg.cmake:748]: this is not a `TARGET_NAME` this is a `PACKAGE_NAME`. The CMake docs always uses `<PackageName>` to describe the variable passed via `find_package`

**Comment by Neumann-A** [comment]
I would really like to see some cleanup first instead of more things added to `vcpkg.cmake`.

**Review by ras0219-msft — COMMENTED** [review]
The pieces I've reviewed LGTM so far.

Still to review:
- New macro/functions
- `find_package()` override

**Comment by strega-nil** [comment]
Depends on #16130

File: scripts/buildsystems/vcpkg.cmake [hunk 5]
```diff
@@ -84,16 +99,215 @@ Install the dependencies listed in your manifest:
...
+        find_program(Z_VCPKG_PWSH_PATH pwsh)
+        if(Z_VCPKG_PWSH_PATH)
+            set(Z_VCPKG_POWERSHELL_PATH "${Z_VCPKG_PWSH_PATH}" CACHE INTERNAL "The path to the PowerShell implementation to use.")
+        else()
+            message(DEBUG "vcpkg: Could not find PowerShell Core; falling back to PowerShell")
+            find_program(Z_VCPKG_BUILTIN_POWERSHELL_PATH powershell REQUIRED)
+            if(Z_VCPKG_BUILTIN_POWERSHELL_PATH)
+                set(Z_VCPKG_POWERSHELL_PATH "${Z_VCPKG_BUILTIN_POWERSHELL_PATH}" CACHE INTERNAL "The path to the PowerShell implementation to use.")
+            else()
+                message(WARNING "vcpkg: Could not find PowerShell; using static string 'powershell.exe'")
```
Comment (BillyONeal) [inline scripts/buildsystems/vcpkg.cmake:293]: I know this was here before but doesn't `REQUIRED` on 289 make this else dead?

File: scripts/buildsystems/vcpkg.cmake [hunk 6]
```diff
@@ -581,28 +784,35 @@ if(X_VCPKG_APPLOCAL_DEPS_INSTALL)
...
-    string(TOLOWER "${name}" _vcpkg_lowercase_name)
+    z_vcpkg_function_arguments(ARGS)
+    set(PACKAGE_NAME "${ARGV0}")
+    string(TOLOWER "${PACKAGE_NAME}" LOWERCASE_PACKAGE_NAME)
 
-    if(EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake")
-        set(ARGS "${ARGV}")
-        include(${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake)
-    elseif("${name}" STREQUAL "Boost" AND EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include/boost")
+    set(VCPKG_CMAKE_WRAPPER_PATH "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${LOWERCASE_PACKAGE_NAME}/vcpkg-cmake-wrapper.cmake")
```
Comment (BillyONeal) [inline scripts/buildsystems/vcpkg.cmake:810]: `${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}` is repeated a lot, should we introduce a `Z_VCPKG_CURRENT_INSTALLED_DIR` ?

File: scripts/buildsystems/vcpkg.cmake [hunk 7]
```diff
@@ -614,79 +824,77 @@ macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
...
+    z_vcpkg_complete_parent_scope_export(IGNORE_REGEX "(^Z_VCPKG_)|(^ARGS$)|(^COMPONENTS_IDX$)")
+endfunction()
 
 set(VCPKG_TOOLCHAIN ON)
-set(_UNUSED ${CMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION})
-set(_UNUSED ${CMAKE_EXPORT_NO_PACKAGE_REGISTRY})
-set(_UNUSED ${CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY})
-set(_UNUSED ${CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY})
-set(_UNUSED ${CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP})
+set(Z_VCPKG_UNUSED "${CMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION}")
```
Comment (BillyONeal) [inline scripts/buildsystems/vcpkg.cmake:867]: Stupid ?: Why are we scribbling over a variable with things then throwing it away?

File: scripts/buildsystems/vcpkg.cmake [hunk 8]
```diff
@@ -614,79 +824,77 @@ macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
...
+set(Z_VCPKG_UNUSED "${CMAKE_EXPORT_NO_PACKAGE_REGISTRY}")
+set(Z_VCPKG_UNUSED "${CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY}")
+set(Z_VCPKG_UNUSED "${CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY}")
+set(Z_VCPKG_UNUSED "${CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP}")
```
Comment (BillyONeal) [inline scripts/buildsystems/vcpkg.cmake:872]: Should there be an `unset(Z_VCPKG_UNUSED)` here?

**Review by BillyONeal — APPROVED** [review]

File: scripts/buildsystems/vcpkg.cmake (hunk 7 above)
Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:867]: These are equivalent to casts to `void`.

File: scripts/buildsystems/vcpkg.cmake (hunk 8 above)
Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:872]: Not really, it's unimportant.

File: scripts/buildsystems/vcpkg.cmake (hunk 6 above)
Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:810]: Yeah probably, but we'll do that in the future.

File: scripts/buildsystems/vcpkg.cmake (hunk 5 above)
Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:293]: Yes, it does. We should remove it in the future.

File: scripts/buildsystems/vcpkg.cmake (hunk 7 above)
Comment (BillyONeal) [reply scripts/buildsystems/vcpkg.cmake:867]: I see ,thanks!

**Comment by cenit** [comment]
@strega-nil Did you try the vcpkg.cmake script after the work on a downstream project?
It looks like it broke vcpkg totally. No packages can be found. I can reproduce it locally and it is being triggered right now in every ci pipeline of mine.
A message is also just appeared about that: https://github.com/microsoft/vcpkg/issues/15956#issuecomment-781859104

**Pull request merged**